sdk/core/azure-core/src/io/body_stream.cpp

```diff
@@ -190,3 +190,32 @@ size_t FileBodyStream::OnRead(uint8_t* buffer, size_t count, Azure::Core::Contex
 void FileBodyStream::Rewind() { m_randomAccessFileBodyStream->Rewind(); }
 
 int64_t FileBodyStream::Length() const { return m_randomAccessFileBodyStream->Length(); }
+
+ProgressBodyStream::ProgressBodyStream(
+    BodyStream& bodyStream,
+    std::function<void(int64_t bytesTransferred)> callback)
+    : m_bodyStream(bodyStream), m_bytesTransferred(0), m_callback(std::move(callback))
+{
+  m_callback(m_bytesTransferred);
+}
+
+void ProgressBodyStream::Rewind()
+{
+  m_bodyStream.Rewind();
+  m_bytesTransferred = 0;
+  m_callback(m_bytesTransferred);
+}
+
+size_t ProgressBodyStream::OnRead(
+    uint8_t* buffer,
+    size_t count,
+    Azure::Core::Context const& context)
+{
+  size_t read = m_bodyStream.Read(buffer, count, context);
```

Comment:
I wonder what happens if I pass in a `ProgressBodyStream` into itself. We may want to add a test case for that :)

Reply:
nothing glorious , the only interesting thing is that both callbacks will be called, as expected, since the progress stream relies on the underling stream functionality to perform any actions.